### ISSUE TITLE
Question about put response time

### ISSUE THREAD

**Issue by kylebernhardy** [DoctorEvidence/lmdb-store#34]
Hi,

We are typically executing many individual async transactions using ifNoExists & in benchmarking we are seeing lower performance than expected. I added some timing around the calls and found that each ifNoExists call is taking around 7-15ms.  I would have thought they would take around 1ms based on the commitDelay rather than the 7-15ms.  Is this expected behavior for individual ifNoExists?

The sample code below isn't a one for one of our code, but meant to show what I'm seeing.  It does look like if I await when multiple async calls are in flight the performance improves.

```

const fs = require('fs');
const path = require('path');
const lmdb = require('lmdb-store');

function createEnv(name){
    let environment_path = path.join(__dirname, 'data', name);
    try {
        fs.rmdirSync(environment_path, {recursive:true})
        fs.mkdirSync(environment_path);
    }catch(e){
        console.error(e);
    }
    let env_init = {
        "path": environment_path,
        "mapSize": 1024*1024,
        "maxReaders": 100
    };
    let env = lmdb.open(env_init);
    return env;
}

let env = createEnv('test');

let id_dbi = env.openDB('id', {
    create: true,
    dupSort: false,
    useVersions: false});
let timestamp_dbi = env.openDB('timestamp', {
    create: true,
    dupSort: false,
    useVersions: false});

(async()=>{
    let promise;
    for(let x = 0, length = 100; x < length; x++){
        let o = {id:x, timestamp: Date.now()};
        console.time('put');
        await id_dbi.ifNoExists(x, ()=>{
            id_dbi.put(o.id, o);
            timestamp_dbi.put(o.timestamp, x);
        });
        console.timeEnd('put');
    }
})()

```

**Comment by kriszyp** [comment]
lmdb-store uses a timer-based (setTimeout) queuing for these commits, and setTimeout is... not very accurate. For comparison, if you make a simple function that simply waits for one millisecond:
```
const delay = () => new Promise(resolve => setTimeout(resolve, 1));
```
And then run your test using that delay instead of `ifNoExists`, I get almost identical results (10-15ms per loop). lmdb-store is simply waiting for the timer to queue any other writes before doing the commit, but 99.9% is waiting, not actual processing. The issue isn't technically "slow performance", but rather high latency.

Alternately I also tried changing action to perform 1000 writes on each loop, 
```
    console.time('put');
    for(let x = 0, length = 1000; x < length; x++){
        let o = {id:x, timestamp: Date.now()};
        promise = id_dbi.ifNoExists(x, ()=>{
            id_dbi.put(o.id, o);
            timestamp_dbi.put(o.timestamp, x);
        });
    }
    await promise
    console.timeEnd('put');
```
And each of these take basically about the same amount of time of 10-15ms. The actual processing may actually be closer to around 10-15 **nanoseconds** per operation. In other words, even we a fixed minimum latency, you can scale this to a massive amount of operations (sequential or parallel) without slowing things down significantly.

But lower latency may still be valuable to you. I can easily add an option to use `setImmediate` (perhaps when commitDelay is 0 or -1), instead of `setTimeout`, which makes the queuing calling much more immediate (hence the name :) ). When I tested that, the original test was running at about 0.3-0.4ms per iteration. Would that be helpful?

**Comment by kylebernhardy** [comment]
That makes a lot of sense, thank you for adding the info regarding what is happening with the setTimeout.  Lower latency per request is important for us, but also having the batching adds nice performance on load .  Would your proposed change still allow for transaction batching on the thread, just on a smaller time frame?  If so that sounds like best of both worlds & would be really helpful.

**Comment by kriszyp** [comment]
Ok, ref'ed commit uses setImmediate when commitDelay is zero (or less than 1). And yes, using setImmediate would still allow for transaction batching on the main thread, with smaller time frame. But it is a significantly smaller time frame, so it definitely could result in less batching and lower efficiency, but worth testing and seeing if it works better for your situation. I will do some testing locally too before publishing next release.

**Comment by kylebernhardy** [comment]
Thanks Kris,

I have an artillery test I can run one the change is posted & compare to previous results.  I will let you know how it looks for us.

**Comment by kriszyp** [comment]
After some local testing, I was probably overestimating the cost of shorter batch times, this seems to work fine. Your desire for shorter timeframes is probably common, so I adjusted the default to be using a delay of 0 and using setImmediate for batching. I have the change published in 1.1.0.
